/*LeetCode problem 268
 Missing Number
 Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.
 Example 1:
 Input: nums = [3,0,1]
 Output: 2
 */


package Sorting;

public class LC268 {
    public static void main(String[] args) {
        int[] nums={3,0,1};
        System.out.println(missingNumber(nums));
    }

    static int missingNumber(int[] nums){
        int i=0;
        while (i<nums.length){
            int correct=nums[i];
            if(nums[i]< nums.length && nums[i] != nums[correct]){
                swap(nums, i, correct);
            }
            else{
                i++;
            }
        }

        for (int j=0; j<nums.length; j++){
            if(nums[j] != j){
                return j;
            }
        }
        return nums.length;
    }

    static void swap(int[] arr, int first,int second){
        int temp=arr[first];
        arr[first]=arr[second];
        arr[second]=temp;
    }
}
